### PR TITLE
Addition service persona. Initial SQLite integration for chat histories.

### PR DESCRIPTION
**Pull request by PaulEloffPraekelt**
_**Addition of service persona:**_
Each chat history now starts off with a system message that contains the service persona. This also required updating Onboarding and Survey pipelines. For the time being, I have left the persona out of assessments (which also do not yet require a chat history), because the assessments are supposed to be less dynamic and do not necessarily require a service persona.

_**Integration of SQLite:**_
Chat histories for onboarding and the ANC survey are now managed through a SQLite database.

Due to the changes in chat history management, and the addition of asynchronous database behaviour, we also had to make changes to our APIs and their tests.

### CODE REVIEW

**Comment by PaulEloffPraekelt** [comment]
I think we'll need some more robust logic (and more tests) around the new `PreAssessmentQuestionHistory` table to make sure we use it as intended. But I think if there's close alignment with Engineering and we know exactly all the ways in which Turn is going to call our endpoints, then that can guide exactly what and where we need to test thoroughly. If we write too many tests without that alignment with Eng, then we might write tests that catch edge cases that would never even have happened.

**Comment by PaulEloffPraekelt** [comment]
PS I also just realized the README needs a bit of an update too, but I'll do that in my next PR.

File: src/ai4gd_momconnect_haystack/api.py [hunk 1]
```diff
@@ -59,91 +74,87 @@ def verify_token(authorization: Annotated[str, Header()]):
...
         user_context = extract_onboarding_data_from_response(
-            user_response=request.user_input,
+            user_response=user_input,
             user_context=request.user_context,
             chat_history=chat_history,
         )
     else:
         # If there is no response to the question, the context stays the same
         user_context = request.user_context
+    if intent == "CHITCHAT" and intent_related_response:
```
[the remark this answers is not in the document]
Comment (PaulEloffPraekelt) [reply src/ai4gd_momconnect_haystack/api.py:116]: I actually still have some doubt about whether we really want to be storing `CHITCHAT` data at all, because the purpose of the chat history is to prime the LLM, which `CHITCHAT` wouldn't help with.

@Themba-Gqaza do you think we need to store the chat histories for any other purpose? We'll save flow results for the things we want to record in any case. And we'll still have complete message histories in Turn's data.

**_But to address your question directly:_** We do store user responses and journey questions in the chat history - just a bit above and below the referenced lines.

File: src/ai4gd_momconnect_haystack/api.py [hunk 2]
```diff
@@ -157,59 +168,74 @@ def assessment(request: AssessmentRequest, token: str = Depends(verify_token)):
...
+            messages=chat_history,
+            history_type=request.survey_id,
         )
-        chat_history.append(f"User to System: {request.user_input}")
-        if intent_related_response:
-            chat_history.append(f"System to User: {intent_related_response}")
+        intent, intent_related_response = handle_user_message(last_question, user_input)
     else:
         intent, intent_related_response = "JOURNEY_RESPONSE", ""
+        if not chat_history:
```
[the remark this answers is not in the document]
Comment (PaulEloffPraekelt) [reply src/ai4gd_momconnect_haystack/api.py:234]: This inconsistency is a result of me doubting exactly which checks we want to implement before starting a 'fresh' chat history, and then not implementing it exactly the same everywhere.

But I am happy with us assuming that an empty message from Turn means that we force a fresh start of that journey, i.e. we create a new chat history regardless of whether there already existed one.

Related to this is our chat in Slack about the usage of chat histories during user testing, which I am pretty sure will be solved by such an implementation as well.

File: src/ai4gd_momconnect_haystack/api.py [hunk 3]
```diff
@@ -157,59 +168,74 @@ def assessment(request: AssessmentRequest, token: str = Depends(verify_token)):
...
+    chat_history = await get_or_create_chat_history(
+        user_id=user_id, history_type=request.survey_id
+    )
+
     last_question = chat_history[-1] if chat_history else ""
     if request.user_input:
-        intent, intent_related_response = handle_user_message(
-            last_question, request.user_input
+        chat_history.append(ChatMessage.from_user(text=user_input))
+        await save_chat_history(
```
[the remark this answers is not in the document]
Comment (PaulEloffPraekelt) [reply src/ai4gd_momconnect_haystack/api.py:226]: True.

File: pyproject.toml [hunk 4]
```diff
@@ -12,6 +12,8 @@ dependencies = [
     "fastapi[standard]>=0.115.12",
     "sentry-sdk[fastapi]>=2.29.1",
     "prometheus-fastapi-instrumentator>=7.1.0",
+    "sqlalchemy>=2.0.29",
+    "aiosqlite>=0.20.0",
```
[the remark this answers is not in the document]
Comment (PaulEloffPraekelt) [reply pyproject.toml:16]: Sure! Under `dependencies`? And should I also perhaps move `aiosqlite` to `dev-dependencies` instead?

File: src/ai4gd_momconnect_haystack/doc_store.py [hunk 5]
```diff
@@ -126,17 +127,27 @@ def ingest_content(
...
-                    valid_responses = list(valid_responses.keys())
+                # For DMA and KAB flows
                 doc = Document(
                     content=piece["content"],
                     meta={
                         "flow_id": flow_id,
                         "question_number": piece["question_number"],
                         "content_type": piece["content_type"],
-                        "valid_responses": valid_responses,
+                        "valid_responses_and_scores": json.dumps(
```
[the remark this answers is not in the document]
Comment (PaulEloffPraekelt) [reply src/ai4gd_momconnect_haystack/doc_store.py:148]: Sure, we can read it from JSON. I kind of started treating the document store as a database in my mind, hence I stored everything about our content pieces in there. But on second thought, the purpose of it is really to provide documents to the LLM, and maybe we shouldn't be using it as a "database" for other things - so happy to refactor the code to read scores from the JSON instead, and not have them be stored in the document store.

File: src/ai4gd_momconnect_haystack/main.py [hunk 6]
```diff
@@ -191,41 +211,107 @@ def run_simulation():
...
+        iterations = 0
+        flow_id = AssessmentType.dma_pre_assessment
         user_context["goal"] = "Complete the assessment"
         max_assessment_steps = 10  # Safety break
         question_number = 1
 
         # Simulate Assessment
-        for _ in range(max_assessment_steps):
+        while iterations < max_assessment_steps:
+            iterations += 1
```
[the remark this answers is not in the document]
Comment (PaulEloffPraekelt) [reply src/ai4gd_momconnect_haystack/main.py:225]: I noticed we used different approaches to looping, so for consistency, I just replaced one with the other. But I realize that the `for` loop is cleaner, so happy to change the `while` loops to for `loops` :)

File: pyproject.toml [hunk 7]
```diff
@@ -12,6 +12,8 @@ dependencies = [
     "fastapi[standard]>=0.115.12",
     "sentry-sdk[fastapi]>=2.29.1",
     "prometheus-fastapi-instrumentator>=7.1.0",
+    "sqlalchemy>=2.0.29",
```
[the remark this answers is not in the document]
Comment (rudigiesler) [reply pyproject.toml:15]: I just noted it when looking through the sqlalchemy docs, they say you need the async extras to use async.

File: src/ai4gd_momconnect_haystack/pipelines.py [hunk 8]
```diff
@@ -380,40 +380,53 @@ def create_next_onboarding_question_pipeline() -> Pipeline | None:
...
-    You can reference the existing User Context and Chat History to modify the tonality and/or phrasing,
-    or even add emojis, but DO NOT change the core meaning of the question nor introduce ambiguity. Ensure that the chat flows
+    You can reference the existing User Context and chat history to modify the tonality and/or phrasing,
+    but DO NOT change the core meaning of the question nor introduce ambiguity. Ensure that the chat flows
     smoothly (e.g. the first message in a chat must not start as if there were preceding messages).
 
     JSON Response:
     """
+    chat_template = [
+        ChatMessage.from_system(prompt_template),
```
[the remark this answers is not in the document]
Comment (rudigiesler) [reply src/ai4gd_momconnect_haystack/pipelines.py:425]: Yes, but inside the prompt, we're putting the user messages. Should we instead be putting the message history here, instead of inside the prompt? I'm worried about the user doing a "ignore everything I've said previously and do X", and it working because it gets included in a system level prompt. I would think it would be better to not have any user input at the system level.

[the remark this answers is not in the document]
Comment (PaulEloffPraekelt) [reply src/ai4gd_momconnect_haystack/pipelines.py:425]: Ah, I get you! I actually tried that and couldn't get it to work. The conclusion that I got to at the time was that it was possible in Haystack 1 but not in Haystack 2, because Haystack 2's philosophy is about having stateless pipelines and `explicitly` passing context wherever you need to do so. 

ChatGPT and Gemini both seem to recommend the currently implemented approach.

[the remark this answers is not in the document]
Comment (PaulEloffPraekelt) [reply src/ai4gd_momconnect_haystack/pipelines.py:425]: Looking at the documentation refreshes my memory a bit - I think the core issue I faced was that I couldn't find a way to unpack a list of chats into that template. Notice how, in the documentation, they have strings that reference variables inside their "historic chats". But instead, I want to reference such a passed a variable in a loop that is external to those strings (i.e. external to the template), which creates a ChatMessage using each of the strings inside of that variable.

Perhaps it's something I can try again, but yeah I honestly struggled, and only ended up with answers pointing me in my current direction.

File: src/ai4gd_momconnect_haystack/api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Themba-Gqaza) [reply src/ai4gd_momconnect_haystack/api.py]: @PaulEloffPraekelt  I'm leaning towards saving both the question & the response.  It will also help us analysing the intent classification pipeline.
